### PR TITLE
Improve the recipe function

### PR DESCRIPTION
**Pull request by juraj-hrivnak**
- The `recipeFunction` now uses `SlotMatchResult#getGivenInput` instead of `SlotMatchResult#getRecipeIngredient` to get the inputs.

- Implemented `InputList` with `.getAt()` `.findMarked()` and `findMarkedOrEmpty()` methods to get the mark.

- Changed the `inputs` parameter of the function to `InputList` type.

- `IIngredient` now implements `IMarkable`.

- Renamed both `ItemStackMixin#getMark` & `ItemStackMixin#setMark` methods to `ItemStackMixin#grs$getMark` & `ItemStackMixin#grs$setMark`. `.getMark()` & `.setMark()` are now part of `ItemStackMixinExpansion`.

### CODE REVIEW

File: src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java [hunk 1]
```diff
@@ -49,23 +49,17 @@ public CraftingRecipe(ItemStack output, List<IIngredient> input, @Nullable Closu
...
-            }
-        }
-        return result;
+        ItemStack output = this.output.copy();
+
+        if (recipeFunction == null || input.isEmpty()) return output;
+
+        InputList inputs = getMatchingList(inv).stream()
+                .map(SlotMatchResult::getGivenInput)
+                .collect(toCollection(InputList::new));
```
Comment (WaitingIdly) [inline src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java:58]: can you switch this back to a for loop?

File: src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java [hunk 2]
```diff
@@ -155,14 +149,45 @@ public IIngredient getRecipeIngredient() {
...
 
         public int getSlotIndex() {
             return slotIndex;
         }
     }
 
+    public static class InputList extends ArrayList<ItemStack> {
+
+        // groovy [] operator
+        public ItemStack getAt(String mark) {
```
Comment (WaitingIdly) [inline src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java:167]: should this also be annotated nullable?

File: src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java [hunk 3]
```diff
@@ -155,14 +149,45 @@ public IIngredient getRecipeIngredient() {
         }
 
         public ItemStack getGivenInput() {
-            return givenInput;
+            // Copy mark from recipeIngredient to givenInput
+            ItemStackMixinExpansion copy = ItemStackMixinExpansion.of(givenInput);
```
Comment (WaitingIdly) [inline src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java:153]: uh, this isnt actually a copy - `ItemStackMixinExpansion#of` is just `(ItemStackMixinExpansion) (Object) stack`. did you mean to run `ItemStackMixinExpansion#exactCopy`? is making a copy even needed here?

File: examples/postInit/vanilla.groovy [hunk 4]
```diff
@@ -110,6 +110,17 @@ crafting.shapedBuilder()
     .replaceByName()
     .register()
 
+// recipeFunction example
+
+crafting.shapelessBuilder()
+    .output(item('minecraft:wooden_shovel'))
+    .input(ore('logWood').reuse(),item('minecraft:wooden_shovel:*').mark('tool'))
+    .recipeFunction { output, inputs, info ->
+        def item = inputs["tool"]
```
Comment (WaitingIdly) [inline examples/postInit/vanilla.groovy:119]: can you replace the double quotes with single quotes and can you comment this to describe what it does (get the marked itemstack)

**Review by WaitingIdly — COMMENTED** [review]
`vanilla.groovy` got moved to a subfolder on the main branch, which is why yours is conflicting

File: src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java (hunk 2 above)
Comment (juraj-hrivnak) [reply src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java:167]: Probably yes

File: src/main/java/com/cleanroommc/groovyscript/api/IIngredient.java [hunk 5]
```diff
@@ -95,13 +98,26 @@ public ItemStack[] getMatchingStacks() {
...
+        @Nullable
+        @Override
+        public String getMark() {
+            return mark;
+        }
+
+        @Override
+        public void setMark(String mark) {
+            this.mark = mark;
+        }
```
Comment (brachy84) [inline src/main/java/com/cleanroommc/groovyscript/api/IIngredient.java:111]: EMPTY should be immutable

File: src/main/java/com/cleanroommc/groovyscript/api/IIngredient.java [hunk 6]
```diff
@@ -140,5 +156,16 @@ public void setAmount(int amount) {
...
```
Comment (brachy84) [inline src/main/java/com/cleanroommc/groovyscript/api/IIngredient.java:169]: ANY should be immutable

File: src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java (hunk 3 above)
Comment (juraj-hrivnak) [reply src/main/java/com/cleanroommc/groovyscript/compat/vanilla/CraftingRecipe.java:153]: >  is making a copy even needed here?

I don't think so.
Renamed it to `itemStack`.

**Review by brachy84 — CHANGES_REQUESTED** [review]
In this pr you assume that the `givenInput` can have a mark, which is not true. The given input is the same item from the current crafting grid, which is never marked. The mark only exists on the input ingredients of the recipe and not the actual matched items. So your example shouldnt even work

**Comment by brachy84** [comment]
Nevermind what i said

**Review by brachy84 — APPROVED** [review]

**Review by WaitingIdly — APPROVED** [review]

**Pull request merged**